Index: components/poly-cmapi-datafeed/poly-cmapi-datafeed.html
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>windows-1252
===================================================================
--- components/poly-cmapi-datafeed/poly-cmapi-datafeed.html	(revision Local version)
+++ components/poly-cmapi-datafeed/poly-cmapi-datafeed.html	(revision Shelved version)
@@ -1,5 +1,3 @@
-
-
 <!-- Imports polymer -->
 <link rel="import" href="../../bower_components/polymer/polymer.html">
 
@@ -26,13 +24,13 @@
         </div>
     </template>
 
-    <!-- CMAPI v.1.3 script and schema -->
-    <script src="../../libs/cmapi/dist/tv4.js"></script>
-    <script src="../../libs/cmapi/dist/cmapi.schema.js"></script>
-    <script src="../../libs/cmapi/dist/cmajs.js"></script>
-    <script src="../../libs/cmapi/dist/cmapi.examples.js"></script>
-    <script src="../../bower_components/httpclient/dist/HTTPClient.js"></script>
+    <!-- CMAPI v.1.3 script and schema TODO: if the MAP includes de CMAPI, then no other component should include it...-->
 
+    <!--script src="../../libs/cmapi/dist/tv4.js"></script-->
+    <!--script src="../../libs/cmapi/dist/cmapi.schema.js"></script-->
+    <!--script src="../../libs/cmapi/dist/cmajs.js"></script-->
+    <!--script src="../../libs/cmapi/dist/cmapi.examples.js"></script-->
+
     <!-- Registers custom element -->
     <script>
         "use strict";
@@ -162,6 +160,12 @@
                 var PATTERN_RECORD='\n';
                 var PATTERN_RECORD_DATA=':';
 
+                if(window.cmajs!=undefined) {
+                    this._cmajs=window.cmajs;
+                }else {
+                    console.debug('CMAJS not detected yet.');
+                }
+
                 //clean up initial header: the total of records...
                 var data=e.detail.response.toString().split(PATTERN_TOTAL);
                 data=data[1];//discard the total and version header...
@@ -191,7 +195,7 @@
                     var index=this._exists(aircraft[counter]);
                     if(index>=0) {//ya existe, no hace falta anadirlo sino actualizar los atributos del avion
                         this._data_feed[index].attributes=aircraft[counter].attributes;
-                        if(window.cmapi===undefined)  {//if we are in standalone mode, do nothing...
+                        if(this._cmajs.runtimes===undefined)  {//if we are in standalone mode, do nothing...
 
                         }else { //if we are in integrated mode, then send the cmapi message
                             this._CMAPI_map_feature_update(aircraft[counter]);
@@ -205,9 +209,10 @@
 
                     //SEND A CMAPI FEATURE ADD!!!!
 
-                    if(window.cmapi===undefined)  {//if we are in standalone mode, do nothing...
+                    if(this._cmajs.runtimes===undefined)  {//if we are in standalone mode, do nothing...
 
                     }else { //if we are in integrated mode, then send the cmapi message
+                        console.debug(this._cmajs);
                         this._CMAPI_map_feature_plot(aircraft[counter]);
                     }
 
@@ -248,7 +253,7 @@
                 var payload={
                     //TODO: THE LAYER_ID MUST BE A CONFIGURABLE PARAMETER OF THE DATAFEED COMPONENT. ALSO, THE OL3MAP SHOULD DEAL WITH THE PROBLEM WHEN THE LAYER DOES NOT EXIST YET. IT WILL BE AUTOMATICALLY CREATED.
                     "overlayId":this.layer_id,
-                    "featureId":cmajs.utils.getUUID(), //WHEN CREATING A FEATURE, THE ID SHOULD BE RANDOM AND UNIQUE?
+                    "featureId":this._cmajs.utils.getUUID(), //WHEN CREATING A FEATURE, THE ID SHOULD BE RANDOM AND UNIQUE?
                     "format":this.format,                 //TODO: FORMAT FOR FEATURES SHOULD BE CONFIGURABLE AND ATTRIBUTE OF THE COMPONENT. GEOJSON AS THE DEFAULT.
                     "crs": {
                         "type": 'name',
@@ -273,7 +278,7 @@
                     message: payload
                 };
                 console.debug(JSON.stringify(plotmsg));
-                if(this._cmajs!=undefined)
+
                 var bol=this._cmajs.runtimes.browser.pubSub.publish(plotmsg);
 
             },
\ No newline at end of file
Index: components/poly-cmapi-ol3/ol3-cmapi-wrapper.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- components/poly-cmapi-ol3/ol3-cmapi-wrapper.js	(revision Local version)
+++ components/poly-cmapi-ol3/ol3-cmapi-wrapper.js	(revision Shelved version)
@@ -40,7 +40,7 @@
         }); 
     }, 
   
-    overlays: {}, 
+    overlays: [],
 
   
     // ======================================================================== 
@@ -73,12 +73,12 @@
             if (payload.hasOwnProperty('overlayId')) { 
                 id = payload.overlayId; 
                 // if overlay already exists, exit 
-                if (ol3_cmapi_wrapper.overlays.hasOwnProperty(id)) {
+                if (this.overlays[id]!=undefined) {
                     console.log('overlay with id ' + id + 'already exists'); 
                     return; 
                 } 
             } else { 
-                id = ol3_cmapi_wrapper.createGuid();
+                id = this.createGuid();
             } 
   
             // get name or use id as name 
@@ -92,8 +92,8 @@
             if (payload.hasOwnProperty('parentId')) { 
                 parentId = payload.parentId; 
                 // create parent if it does not exist 
-                if (!ol3_cmapi_wrapper.overlays.hasOwnProperty(parentId)) {
-                    ol3_cmapi_wrapper.overlay.create("map.overlay.create", {
+                if (!this.overlays.hasOwnProperty(parentId)) {
+                    this.overlay.create("map.overlay.create", {
                         overlayId: parentId 
                     }); 
                 } 
@@ -102,21 +102,28 @@
                 parentId = undefined; 
             } 
   
-            // now create the overlay and add to map 
+            // now create the overlay and add to map
-            newOverlay = L.featureGroup(); 
-            newOverlay.addTo(ol3_map);
-            ol3_map.getPanes()[id] = ""; //TODO: shoudl remove features before doing this
-
+            console.debug('Creating Empty Layer in OpenLayers 3 API.');
+            var vectorSource = new ol.source.Vector({});
+            console.debug (+vectorSource);
+            var new_Layer = new ol.layer.Vector({'source':vectorSource});
+            console.debug(new_Layer);
+            try {
+                this.ol3_map.addLayer(new_Layer);
+            }catch(err) {
+                console.error(err.message);
+            }
+            console.debug('Added the empty Layer to the OL3 MAP...');
             // add to overlay data store 
-            ol3_cmapi_wrapper.overlays[id] = {
-                obj: newOverlay, 
+            this.overlays[id] = {
+                obj: new_Layer,
                 name: name, 
                 parent: parentId, 
                 children: {}
-            } 
+            };
   
             console.log('Created new overlay with id ' + id); 
-            console.log(ol3_cmapi_wrapper.overlays[id]);
+            console.log(this.overlays[id]);
         }, 
           
         /** 
@@ -142,28 +149,28 @@
             id = payload.overlayId; 
   
             // Case 2: overlay with passed id doesn't exist 
-            if (!ol3_cmapi_wrapper.overlays.hasOwnProperty(id)) {
+            if (!this.overlays.hasOwnProperty(id)) {
                 console.log('no overlay with id ' + id + ' exists'); 
                 return; 
             } 
   
             // Case 3: overlay exists; time to remove it 
-            overlay = ol3_cmapi_wrapper.overlays[id];
+            overlay = this.overlays[id];
             overlayObj = overlay.obj; 
   
             // delete from map; also deletes all child overlays 
             ol3_map.removeLayer(overlayObj);
   
             // delete all children in storage 
-            for (var key in ol3_cmapi_wrapper.feature.features) {
-                var feature = ol3_cmapi_wrapper.feature.features[key].overlayId;
+            for (var key in this.feature.features) {
+                var feature = this.feature.features[key].overlayId;
                 if (feature === id) {
-                    delete ol3_cmapi_wrapper.feature.features[key]
+                    delete this.feature.features[key]
                 }
                 console.log('child overlay with id ' + id + ' removed'); 
             } 
             // and then delete the overlay in storage 
-            delete ol3_cmapi_wrapper.overlays[id];
+            delete this.overlays[id];
   
             console.log('overlay with id ' + id + ' removed'); 
         }, 
@@ -184,7 +191,7 @@
                 return;
             }
 
-            if (!ol3_cmapi_wrapper.overlays.hasOwnProperty(payload.overlayId)) {
+            if (!this.overlays.hasOwnProperty(payload.overlayId)) {
                 console.log("Overlay doesn't exist to hide: " + payload.overlayId);
                 return;
             }
@@ -213,7 +220,7 @@
                 return;
             }
 
-            if (!ol3_cmapi_wrapper.overlays.hasOwnProperty(payload.overlayId)) {
+            if (!this.overlays.hasOwnProperty(payload.overlayId)) {
                 console.log("Overlay doesn't exist to hide: " + payload.overlayId);
                 return;
             }
@@ -240,7 +247,7 @@
                 return;
             } else { // Update attributes
                 //TODO: feature.features needs to be updated too
-                var overlay = ol3_cmapi_wrapper.overlays[payload.overlayId];
+                var overlay = this.overlays[payload.overlayId];
 
                 if (payload.hasOwnProperty("name")) {
                     overlay.name = payload.name;
@@ -249,8 +256,8 @@
                 if (payload.hasOwnProperty("parentId")) {
                     overlay.parent = payload.parentId;
                     // create parent if it does not exist 
-                    if (!ol3_cmapi_wrapper.overlays.hasOwnProperty(payload.parentId)) {
-                        ol3_cmapi_wrapper.overlay.create("map.overlay.create", {
+                    if (!this.overlays.hasOwnProperty(payload.parentId)) {
+                        this.overlay.create("map.overlay.create", {
                             overlayId: payload.parentId 
                         }); 
                     } 
@@ -264,7 +271,7 @@
     // map.feature channels: load feature data onto map 
     feature: { 
 
-        features: {},//TODO: STUDY THE WAY OF ORGNISING THIS OR DELEGATING TO THE MAPPING LIBRARY. ALL FEATURES WOULD BE STORED HERE. EACH FEATURE WILL BE ASOCIATED WITH AN OVERLAY, BUT ALL ARE MIXED HERE. THIS WILL CONTAIN CMAPI STRUCTURES, AND ILL BE REPEATED IN THE MAP ITSELF????
+        features: [],//TODO: STUDY THE WAY OF ORGNISING THIS OR DELEGATING TO THE MAPPING LIBRARY. ALL FEATURES WOULD BE STORED HERE. EACH FEATURE WILL BE ASOCIATED WITH AN OVERLAY, BUT ALL ARE MIXED HERE. THIS WILL CONTAIN CMAPI STRUCTURES, AND ILL BE REPEATED IN THE MAP ITSELF????
   
         // Channel names for map.feature 
         channels: {
@@ -296,26 +303,113 @@
         plotFeature: function (channel, payload) { 
             console.log('map.feature.plot triggered'); 
             var overlayId;
-
+            console.log(this.ol3_map);
             // if (_.isEmpty(payload.featureId) || _.isEmpty(payload.feature)) {
             //     console.log("FeatureID & Feature attributes are required");
             // }
 
             // Create overlay id
-            if (_.isEmpty(payload.overlayId) ||
-                _.isEmpty(ol3_cmapi_wrapper.overlays[payload.overlayId])) {
-                overlayId = ol3_cmapi_wrapper.createGuid();
-                ol3_cmapi_wrapper.overlay.create('channel', {overlayId: overlayId});
+            if (!payload.hasOwnProperty('overlayId') ||
+                (this.overlays[payload.overlayId]===undefined)) {
+                console.debug('The Overlay does NOT exist...');
+                overlayId = "AIRCRAFT24"; //TODO: How to transmit back the Overlay ID to the Data Feed??? . this.createGuid();
+                payload.overlayId = overlayId; //ADDED for HACK.
+
+                this.overlay.create(channel, {overlayId: overlayId});
+                var geojsonObject=payload;
+
+
+                var vectorSource=this.overlays[payload.overlayId].obj.getSource();
+                vectorSource.addFeatures((new ol.format.GeoJSON()).readFeatures(geojsonObject));
+
+                var styles = {
+                    'Point': [new ol.style.Style({
+                        image: image
+                    })],
+                    'LineString': [new ol.style.Style({
+                        stroke: new ol.style.Stroke({
+                            color: 'green',
+                            width: 1
+                        })
+                    })],
+                    'MultiLineString': [new ol.style.Style({
+                        stroke: new ol.style.Stroke({
+                            color: 'green',
+                            width: 1
+                        })
+                    })],
+                    'MultiPoint': [new ol.style.Style({
+                        image: image
+                    })],
+                    'MultiPolygon': [new ol.style.Style({
+                        stroke: new ol.style.Stroke({
+                            color: 'yellow',
+                            width: 1
+                        }),
+                        fill: new ol.style.Fill({
+                            color: 'rgba(255, 255, 0, 0.1)'
+                        })
+                    })],
+                    'Polygon': [new ol.style.Style({
+                        stroke: new ol.style.Stroke({
+                            color: 'blue',
+                            lineDash: [4],
+                            width: 3
+                        }),
+                        fill: new ol.style.Fill({
+                            color: 'rgba(0, 0, 255, 0.1)'
+                        })
+                    })],
+                    'GeometryCollection': [new ol.style.Style({
+                        stroke: new ol.style.Stroke({
+                            color: 'magenta',
+                            width: 2
+                        }),
+                        fill: new ol.style.Fill({
+                            color: 'magenta'
+                        }),
+                        image: new ol.style.Circle({
+                            radius: 10,
+                            fill: null,
+                            stroke: new ol.style.Stroke({
+                                color: 'magenta'
+                            })
+                        })
+                    })],
+                    'Circle': [new ol.style.Style({
+                        stroke: new ol.style.Stroke({
+                            color: 'red',
+                            width: 2
+                        }),
+                        fill: new ol.style.Fill({
+                            color: 'rgba(255,0,0,0.2)'
+                        })
+                    })]
+                };
+
+                var styleFunction = function(feature, resolution) {
+                    return styles[feature.getGeometry().getType()];
+                };
+
+                var vectorLayer = new ol.layer.Vector({
+                    source: vectorSource,
+                    style: styleFunction
+                });
+
+                this.ol3_map.addLayer(vectorLayer);
             } else { // Overlay exists.
+                console.debug('The Overlay exists...');
                 overlayId = payload.overlayId;
+                //TODO: Determine if the overlay exists, I don´t know if it is worth to update the source or just recreate from scratch
+
             }
 
-            //THIS CODE CONTAINS LEAFLET STUFF (L, PANE). TODO: ADAPT THE ADD FEATURE CODE TO OL3
-
             if (payload.format === "geojson") {
-                var obj = new L.geoJson(payload.feature, {pane: overlayId});
-                ol3_map.addLayer(obj);
-                ol3_cmapi_wrapper.feature.features[payload.featureId] =
+                console.debug('Format is geoJSON...');
+                var geojsonObject=payload;
+                var vectorSource=this.overlays[payload.overlayId].obj.getSource();
+                vectorSource.addFeatures((new ol.format.GeoJSON()).readFeatures(geojsonObject));
+                this.feature.features[payload.featureId] =
                     {marker: obj,
                         overlayId: overlayId};
                 // TODO: Should check if it exists, if it does remove -> add new one
@@ -343,8 +437,8 @@
         unplotFeature: function (channel, payload) { 
             console.log('map.feature.unplot triggered'); 
             //TODO: Add error checking.
-            ol3_map.removeLay(ol3_cmapi_wrapper.feature.features[payload.featureId].marker);
-            delete ol3_cmapi_wrapper.feature.features[payload.featureId];
+            ol3_map.removeLay(this.feature.features[payload.featureId].marker);
+            delete this.feature.features[payload.featureId];
             console.log('Feature removed'); 
         }, 
           
@@ -358,7 +452,7 @@
         hideFeature: function (channel, payload) { 
             console.log('map.feature.hide triggered'); 
             //TOD: Add error checking
-            ol3_cmapi_wrapper.feature.features[payload.featureId].marker.setOpacity(0);
+            this.feature.features[payload.featureId].marker.setOpacity(0);
             // marker.setOpacity(1 or 0);
         }, 
           
@@ -372,7 +466,7 @@
         showFeature: function (channel, payload) { 
             console.log('map.feature.show triggered'); 
             //TOD: Add error checking
-            ol3_cmapi_wrapper.feature.features[payload.featureId].marker.setOpacity(1);
+            this.feature.features[payload.featureId].marker.setOpacity(1);
             // marker.setOpacity(1 or 0);
         }, 
           
@@ -438,10 +532,10 @@
             console.log('map.view.center.overlay triggered');
             if (typeof payload.overlayId === 'string'){
                 //Take the overlay ID and use it to find the overlay bounds
-                var bounds = ol3_cmapi_wrapper.overlays[payload.overlay].obj.getBounds();
+                var bounds = this.overlays[payload.overlay].obj.getBounds();
             } else {
                 //Find the default overlay and use its center
-                var bounds = ol3_cmapi_wrapper.overlays['default'].obj.getBounds();
+                var bounds = this.overlays['default'].obj.getBounds();
             }
             if (typeof payload.zoom === 'number'){
                 var havezoom = true;
@@ -468,10 +562,10 @@
             console.log('map.view.center.feature triggered');
             if (typeof payload.overlayId === 'string'){
                 //Take the overlay ID and use it to find the overlay bounds
-                var bounds = ol3_cmapi_wrapper.overlays[payload.overlay].obj.getBounds();
+                var bounds = this.overlays[payload.overlay].obj.getBounds();
             } else {
                 //Find the default overlay and use its center
-                var bounds = ol3_cmapi_wrapper.overlays['default'].obj.getBounds();
+                var bounds = this.overlays['default'].obj.getBounds();
            }
             if (typeof payload.zoom === 'number'){
                 var havezoom = true;
@@ -604,7 +698,7 @@
                 switch(payload.types[type]) {
                     case 'view':
                         console.debug('Publishing a map.status.view response for ' + senderID.id);
-                        res_payload=ol3_cmapi_wrapper.status.respondWithMapViewStatus();
+                        res_payload=this.status.respondWithMapViewStatus();
                         //GOT THE PAYLOAD
                         //NOW WRAPPIT, AND PUBLISH
 
@@ -612,19 +706,19 @@
 
                         var msg = {
                             channel: "map.status.view",
-                            message: res_payload,
+                            message: res_payload
                         };
                         //console.debug(JSON.stringify(msg));
                         var bol=cmajs.runtimes.browser.pubSub.publish(msg);
                         //TODO:PERFORM ERROR CONTROL ON SENDING THE RESPONSE TO A MESSAGE.
                         if(!bol) {
                             console.error('An error ocurred sending the map.status.view response to the map.status.request message.');
-                        };
+                        }
 
                         continue;
                     case 'about':
                         console.debug('Publishing a map.status.about response');
-                        res_payload=ol3_cmapi_wrapper.status.respondWithMapAboutStatus();
+                        res_payload=this.status.respondWithMapAboutStatus();
                         //GOT THE PAYLOAD
                         //NOW WRAPPIT, AND PUBLISH
                         //TODO:EXTRACT A TEMPLATE MESSAGE FOR MAP.STATUS.ABOUT, ADD THE CORRECT SENDERID, ADD THE PAYLOAD AND PUBLISH IT.
@@ -632,7 +726,7 @@
 
                     case 'selected':
                         console.debug('Publishing a map.status.selected response');
-                        res_payload=ol3_cmapi_wrapper.status.respondWithMapSelectedStatus();
+                        res_payload=this.status.respondWithMapSelectedStatus();
                         //GOT THE PAYLOAD
                         //NOW WRAPPIT, AND PUBLISH
                         //TODO:EXTRACT A TEMPLATE MESSAGE FOR MAP.STATUS.SELECTED.VIEW, ADD THE CORRECT SENDERID, ADD THE PAYLOAD AND PUBLISH IT.
@@ -640,7 +734,7 @@
 
                     case 'initialization' :
                         console.debug('Publishing a map.status.initialization response');
-                        res_payload=ol3_cmapi_wrapper.status.respondWithMapInitializationStatus();
+                        res_payload=this.status.respondWithMapInitializationStatus();
                         //GOT THE PAYLOAD
                         //NOW WRAPPIT, AND PUBLISH
                         //TODO:EXTRACT A TEMPLATE MESSAGE FOR MAP.STATUS.INITIALIZATION., ADD THE CORRECT SENDERID, ADD THE PAYLOAD AND PUBLISH IT.
@@ -648,7 +742,7 @@
 
                     case 'format':
                         console.debug('Publishing a map.status.format response');
-                        res_payload=ol3_cmapi_wrapper.status.respondWithMapFormatStatus();
+                        res_payload=this.status.respondWithMapFormatStatus();
                         //GOT THE PAYLOAD
                         //NOW WRAPPIT, AND PUBLISH
                         //TODO:EXTRACT A TEMPLATE MESSAGE FOR MAP.STATUS.FORMAT, ADD THE CORRECT SENDERID, ADD THE PAYLOAD AND PUBLISH IT.
@@ -688,7 +782,7 @@
                 this.ol3_map=document.querySelector('#ol3_map')._ol3_map;
                 this.ol3_map.updateSize();
                 var view=this.ol3_map.getView();
-                view.on('propertychange', ol3_cmapi_wrapper.onMapViewChange,ol3_cmapi_wrapper);
+                view.on('propertychange', this.onMapViewChange,ol3_cmapi_wrapper);
             }
             var view=this.ol3_map.getView();
             var projection=view.getProjection();
@@ -697,7 +791,6 @@
             var center_lonlat = ol.proj.transform(center, 'EPSG:3857', 'EPSG:4326');
             var center_lon = center_lonlat[0];
             var center_lat = center_lonlat[1];
-            var projection = view.getProjection();
             var size=this.ol3_map.getSize();
             var extent=view.calculateExtent(size);
             var southWest=[extent[0],extent[1]];
@@ -901,6 +994,7 @@
      ]
    } 
 };
+
 
 //cmapi.feature.plotFeature('channel', geotest);
 //map.setView({lat:102.0, lng:0.5});
